packages/x-data-grid-pro/src/hooks/features/serverSideData/useGridDataSource.ts

```diff
@@ -0,0 +1,288 @@
+import * as React from 'react';
+import {
+  useGridApiEventHandler,
+  gridRowsLoadingSelector,
+  useGridApiMethod,
+  GridServerSideGroupNode,
+  GridRowId,
+  useGridSelector,
+} from '@mui/x-data-grid';
+import { gridRowGroupsToFetchSelector, GridStateInitializer } from '@mui/x-data-grid/internals';
+import { GridGetRowsResponse } from '../../../models';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+import {
+  gridGetRowsParamsSelector,
+  gridServerSideDataLoadingSelector,
+  gridServerSideDataErrorsSelector,
+} from './gridServerSideDataSelector';
+import { GridDataSourceApi, GridDataSourcePrivateApi } from './serverSideInterfaces';
+import { runIfServerMode, NestedDataManager, RequestStatus } from './utils';
+
+const INITIAL_STATE = {
+  loading: {},
+  errors: {},
+};
+
+export const dataSourceStateInitializer: GridStateInitializer = (state, _, apiRef) => {
+  apiRef.current.caches.serverSideData = {
+    groupKeys: [],
+  };
+
+  return {
+    ...state,
+    serverSideData: INITIAL_STATE,
+  };
+};
+
+export const useGridDataSource = (
```

Comment:
Something I don't like about our initialization structure is that everyone pays the cost for the methods/effects allocations for server-side code even if they don't use those features. Not sure how to solve it though.

Reply:
That's a good point, seems like a more generic one though. It'd be nice if we solved it at some point.